Adds Emag functionality to Cleanbots & Floorbots

AtnerNT opened the pull request:
### Purpose
Adds the ability to emag Cleanbots (Both variants) and Floorbots, aswell as providing a minor bugfix to Floorbots that would stop them from fixing floor tiles that had wires or pipes running underneath them.

### Notes
Emagged bots will have their sensors turn from green to red indicating sabotage.
Emagged Floorbots will seek out floor tiles and rip them up.
Emagged Cleanbots will run around spilling lube on tiles.

AtnerNT committed:
Emaggable Clean and Floor Bots

Allows for traitors to emag floor and clean bots. Emagging the bots gives a small visual change and changes the bots AI. Floorbots will rip up floor tiles and Cleanbots will begin cleaning every tile in sight, creating messes of wet floor wherever they go. (Also includes a minor bug fix for floorbots)

AtnerNT committed:
Fixes Emag Bots

corp-0 commented on UnityProject/Assets/Scripts/Robotics/Bots/InteractableBot.cs:
```diff
@@ -0,0 +1,58 @@
+	{
+		HandApply interaction;
+
+		[SerializeField] private SpriteRenderer SPRITE_RENDERER;
```
should use ``SpriteHandler`` instead of ``SpriteRenderer`` directly. The former has the advantage of networking the sprite changes. As it is right now, sprites won't change on all connected clients.

corp-0 commented on UnityProject/Assets/Scripts/Robotics/Bots/InteractableBot.cs:
```diff
@@ -0,0 +1,58 @@
+				TryEmag(emag, interaction);
+			}
+		}
+		public void TryEmag(Emag emag, HandApply interaction)
```
the name of this method suggests it will attempt to do something that could fail or that it can ``out`` a reference and returns a bool.

corp-0 commented on UnityProject/Assets/Scripts/Robotics/Bots/InteractableBot.cs:
```diff
@@ -0,0 +1,58 @@
+				return mobController;
+			}
+		}
+		public void ClientPredictInteraction(HandApply interaction) { }
```
if you're not going to implement prediction, use the interaction interface that has no prediction

corp-0 commented on UnityProject/Assets/Scripts/Robotics/Bots/InteractableBot.cs:
```diff
@@ -0,0 +1,58 @@
+		}
+		public void ServerPerformInteraction(HandApply interaction)
+		{
+			this.interaction = interaction;
```
what is this? Was it part of the guide?

AtnerNT replied:
Something that was once relavant that I forgot to delete

AtnerNT replied:
Will get rid of when I submit SpriteHandler changes

corp-0 reviewed (changes requested):
Some changes are required, the most important I see is converting your code to use ``SpriteHandler``. Ask around in discord if you don't know how to set that up.

AtnerNT committed:
Switches to Sprite Handler

Switches from using SpriteRenderer to Sprite handler in InteractableBot.cs as well as updating the relevant prefabs. Renames TryEmag hook and switches from Predicted to normal interable interface.

corp-0 commented on UnityProject/Assets/Scripts/Robotics/Bots/InteractableBot.cs:
```diff
@@ -0,0 +1,52 @@
+	public class InteractableBot : NetworkBehaviour, ICheckedInteractable<HandApply>
+	{
+
+		[SerializeField] private SpriteHandler SPRITE_HANDLER;
```
```suggestion
		[SerializeField] private SpriteHandler spriteHandler;
```

AtnerNT committed:
Update UnityProject/Assets/Scripts/Robotics/Bots/InteractableBot.cs

Co-authored-by: Gilles <43683714+corp-0@users.noreply.github.com>

corp-0 commented on UnityProject/Assets/Scripts/Robotics/Bots/InteractableBot.cs:
```diff
@@ -0,0 +1,52 @@
+
+			return MobController != null;
+		}
+		public void ServerPerformInteraction(HandApply interaction)
```
separate methods with a new line

corp-0 commented on UnityProject/Assets/Scripts/Robotics/Bots/InteractableBot.cs:
```diff
@@ -0,0 +1,52 @@
+		{
+			if (MobController == null) return;
+			MobController.IsEmagged = true;
+			if (EMAGGED_SPRITE != null && SPRITE_HANDLER != null) SPRITE_HANDLER.SetSprite(EMAGGED_SPRITE);
```
does this work? I remember SpriteHandler used indexes instead of sprites.

AtnerNT replied:
I've run it and it works fine

corp-0 replied:
aight, good

corp-0 commented on UnityProject/Assets/Scripts/NPC/AI/MobExplore.cs:
```diff
@@ -28,6 +29,9 @@ public enum Target
 		}
 
+		[Tooltip("The reagent used by emagged cleanbots")]
+		[SerializeField] private Reagent CB_REAGENT;
```
```suggestion
		[SerializeField] private Reagent cbReagent;
```
Style for privates is camelCase

AtnerNT replied:
I thought constant-like variables in development guidelines were asked to be in all caps like normal constants.

corp-0 replied:
uhmmm, the rest of the code handles serialized fields just like any other private, even when they are cashing a reference.

AtnerNT replied:
(Replied to wrong thing). Um the CB_REAGENT here is never changed and always stays the same value so I treated it as a "Constant-Like" variable

AtnerNT replied:
The thing I'm refering to is the comment on No.6 in development standards. "this is not an actual constant, but is intialized in Start() and never modified afterwards so it should still be named using this convention"

corp-0 replied:
it is not a serializable field tho. It is fine, leave it as is. We should eventually move it somewhere else anyways because that reference is only relevant for the cleanbot from all the mobs that use MobExplore.
I really want to rework the way we do AI.

AtnerNT replied:
Yeah, working with AI is a bit annoying as all bots and some mobs AI all come from the same script with the only thing distinguishing their behaviour is there being a couple switches and/or ifs

AtnerNT committed:
Very Minor Code Style changes

Renames SPRITE_HANDLER to spriteHandler and seperates some methods with lines that weren't there before.

AtnerNT committed:
Merge branch 'botemag' of https://github.com/AtnerNT/unitystation into botemag
